lib/utils.js

```diff
@@ -0,0 +1,63 @@
+const mongoUtils = require('@nexes/mongo-utils');
+const nqlLang = require('@nexes/nql-lang');
+const cloneDeep = require('lodash.clonedeep');
+
+/**
+ * ## Merge Filters
+ * Util to combine overrides, default and custom filters based on
+ * following hierarchy: overrides -> custom -> defaults
+ *
+ * overrides - filters which must ALWAYS be applied
+ * defaults - filters which must be applied if a matching filter isn't provided
+ * custom - custom filters which are additional
+ *
+ * @TODO:
+ *
+ * - make function signature more generic to accept any amount of arguments/filters
+ * - and then move the function to mongo-utils?
+ */
+module.exports.mergeFilters = ({overrides, defaults, custom} = {}) => {
+    if (custom && !overrides && !defaults) {
+        return custom;
+    }
+
+    let merged = {};
+
+    if (overrides) {
+        merged = overrides;
+    }
+
+    if (custom && Object.keys(custom).length > 0) {
+        custom = mongoUtils.rejectStatements(custom, (statement) => {
+            return mongoUtils.findStatement(merged, statement);
+        });
+
+        if (custom) {
+            merged = merged ? mongoUtils.combineFilters(merged, custom) : custom;
+        }
+    }
+
+    if (defaults && Object.keys(defaults).length > 0) {
+        defaults = mongoUtils.rejectStatements(defaults, (statement) => {
+            return mongoUtils.findStatement(merged, statement);
+        });
+
+        if (defaults) {
+            merged = merged ? mongoUtils.combineFilters(merged, defaults) : defaults;
+        }
+    }
+
+    return merged;
+};
+
+module.exports.expandFilters = (mongoJSON, expansions) => {
+    const expansionsClone = cloneDeep(expansions);
```

Comment:
@kirrg001 I'm wandering :thinking:  if we can avoid cloning here? the reason for it is that we don't do `nqlLang.parse(expansion.expansion)` twice on the same expansion, right? Instead, we could check if `expansion.expansion` is a string or an object and parse it when it's not an object?

Reply:
IMO the reason is that we modify the original passed options. because its a ref

Reply:
I see the reason. I'm just trying to think if there's an upside of dividing this method into two: 1. parsing expansions 2. applying the utility
It's a lot easier to test it this way, and the method then just does really one thing :)

Reply:
i don't mind - up to you :) , but how does dividing the methods avoids using clone? 🤔

Reply:
My initial thinking about clone was wrong. We need to clone as we'd mess up user input, you are right :)
But dividing it into 2 methods makes testing a lot easier, so I'll just to that.

Reply:
👍 One thing we can do

- devide into two fns
- receive options
- create a new object
- iterate over original expansions
- push the atomic values into the new object
- return new object

This would i think avoid cloning.

Reply:
Yes, this sounds great. No need to have lodash :chains:  dependency then! :tada:

Reply:
🙊